Use Stream's BeginRead/WriteInternal from FileStream

stephentoub opened the pull request:
FileStream has two modes, decided upon at construction time.  When it's created in non-async mode, the Read/WriteAsync methods end up queueing work items to invoke the synchronous Read/Write methods.  To do this, the base methods on Stream delegate to Begin/EndRead/Write (since they were around first) and then the resulting IAsyncResult is wrapped.  However, Stream has an optimization that checks to see whether the derived stream actually overrides Begin/EndXx, and if it doesn't, then it skips using those and goes straight to queueing a work item to Read/Write.  However, FileStream does override those, but when it delegates to the base implementation because it's in non-async mode (rather than because it's a type derived from FileStream), going through Begin/EndXx is just unnecessary overhead.  So, in the right circumstances, we can call to Stream's special helper instead.

cc: @JeremyKuhne 
Fixes https://github.com/dotnet/corefx/issues/42419

|   Method |        Toolchain |     Mean |    Error |   StdDev | Ratio | RatioSD |    Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |----------------- |---------:|---------:|---------:|------:|--------:|---------:|------:|------:|----------:|
|  ReadAll | \new\corerun.exe | 32.44 ms | 0.312 ms | 0.292 ms |  0.94 |    0.01 | 125.0000 |     - |     - |   1.09 MB |
|  ReadAll | \old\corerun.exe | 34.64 ms | 0.248 ms | 0.232 ms |  1.00 |    0.00 | 266.6667 |     - |     - |   1.95 MB |
|          |                  |          |          |          |       |         |          |       |       |           |
| WriteAll | \new\corerun.exe | 33.52 ms | 0.465 ms | 0.435 ms |  0.96 |    0.02 | 125.0000 |     - |     - |   1.09 MB |
| WriteAll | \old\corerun.exe | 34.87 ms | 0.259 ms | 0.216 ms |  1.00 |    0.00 | 266.6667 |     - |     - |   1.95 MB |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System;
using System.IO;
using System.Threading.Tasks;

[MemoryDiagnoser]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    private string _path;
    private Stream _stream;

    private const int NumSegments = 10_240;

    private static byte[] _buffer = new byte[1024];

    [GlobalSetup]
    public void Setup()
    {
        new Random(42).NextBytes(_buffer);

        _path = Path.GetTempFileName();
        using (var fs = File.OpenWrite(_path))
        {
            for (int i = 0; i < NumSegments; i++)
            {
                fs.Write(_buffer, 0, _buffer.Length);
            }
        }

        _stream = new FileStream(_path, FileMode.Open, FileAccess.ReadWrite, FileShare.None, 1, useAsync: false);
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _stream.Dispose();
        File.Delete(_path);
    }

    [Benchmark]
    public async Task ReadAll()
    {
        _stream.Position = 0;
        while (await _stream.ReadAsync(_buffer, 0, _buffer.Length) != 0) ;
    }

    [Benchmark]
    public async Task WriteAll()
    {
        _stream.Position = 0;
        for (int i = 0; i < NumSegments; i++)
        {
            await _stream.WriteAsync(_buffer, 0, _buffer.Length);
        }
    }
}
```

JeremyKuhne reviewed (approved):
Nice! Thanks, @stephentoub!

poizan42 commented on src/System.Private.CoreLib/shared/System/IO/FileStream.cs:
```diff
@@ -372,6 +381,17 @@ public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken
+                // rather than something derived from it and what our BeginRead implementation is going to do.
+                return MemoryMarshal.TryGetArray(buffer, out ArraySegment<byte> segment) ?
+                    new ValueTask<int>((Task<int>)base.BeginReadInternal(segment.Array!, segment.Offset, segment.Count, null, null, serializeAsynchronously: true, apm: false)) :
+                    base.ReadAsync(buffer, cancellationToken);
```
Wouldn't we also want to use BeginReadInternal in the non-array case? The base implementation https://github.com/dotnet/coreclr/blob/3834f562e60a22e9f7feb4f12e05cf5e841257e6/src/System.Private.CoreLib/shared/System/IO/Stream.cs#L369 uses an array from ArrayPool, might need to duplicate the same logic here?

stephentoub replied:
> Wouldn't we also want to use BeginReadInternal in the non-array case?

It's not worth all the logic that would need to be duplicated changed.

poizan42 replied:
That was also my thought when I looked into it, but arguably that overload exists for performance reasons when you need to read into a buffer that isn't a managed array, so might be a bit important to get the maximal performance here (?)

stephentoub replied:
A queued work item per operation is a much larger cost. If you care about "maximal perf", you shouldn't be using the async APIs with a stream not configured for it.

stephentoub committed:
Use Stream's BeginReadInternal from FileStream

FileStream has two modes, decided upon at construction time.  When it's created in non-async mode, the Read/WriteAsync methods end up queueing work items to invoke the synchronous Read/Write methods.  To do this, the base methods on Stream delegate to Begin/EndRead/Write (since they were around first) and then the resulting IAsyncResult is wrapped.  However, Stream has an optimization that checks to see whether the derived stream actually overrides Begin/EndXx, and if it doesn't, then it skips using those and goes straight to queueing a work item to Read/Write.  However, FileStream does override those, but when it delegates to the base implementation because it's in non-async mode (rather than because it's a type derived from FileStream), going through Begin/EndXx is just unnecessary overhead.  So, in the right circumstances, we can call to Stream's special helper instead.
